### ISSUE TITLE
Using volumes flag for mounting causes issue:  Root folder could not be created. Path: /rootBucket in v2.5.0 and v2.5.1

### ISSUE THREAD

**Issue by JoeNonExotic** [adobe/S3Mock#786]
#### TL;DR
* Unable to use volume mount with releases `2.5.0` and `2.5.1`

#### Here is my simple setup: 
```
.
├── docker-compose.yaml
└── tobemounted
    └── test.yaml

1 directory, 2 files
```

#### Here are the contents of my `docker-compose.yaml`: 

```
version: "3.9"
services:
  s3mock:
    image: adobe/s3mock
    environment:
      initialBuckets: testbucket
      root: rootBucket
    volumes:
      - ./tobemounted:/rootBucket/testbucket/
```

#### Expected
* Container spins up as expected 

#### Actual
```
test-s3mock-1  | Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [java.io.File]: Factory method 'rootFolder' threw exception; nested exception is java.lang.IllegalStateException: Root folder could not be created. Path: /rootBucket
```

### Additional Info
* Interestingly this seems to be happening only with `2.5.x` releases and not with `2.4.x`, for example if i alter my `docker-compose.yaml` to look like below, it works as expected 

```
version: "3.9"
services:
  s3mock:
    image: adobe/s3mock:2.4.16
    environment:
      initialBuckets: testbucket
      root: rootBucket
    volumes:
      - ./tobemounted:/rootBucket/testbucket/
```

It _seems_ to be a valid issue, but I am also new to this repo, so I may be doing something horrendously wrong as well. Glad to share more info (ex: enable debug/traces and provide with output etc).

**Comment by afranken** [comment]
@JoeNonExotic thanks for the report! :)
When refactoring for 2.5.x, I tried to forward exceptions to the application, log them etc.
Many times, when files are created or deleted, S3Mock would previously ignore exceptions or return values (e.g., if a directory could be created or deleted successfully).

For the root folder, I made a mistake:
Usually, the folder would not exist, so it needs to be created.
In your case (which we actually do not test in S3Mock), you mount a volume, and specify the mounted folder as the root for S3Mock.
I rewrote the code so that this is interpreted as an error, even though it's a feature of the application, D'OH!

**Comment by afranken** [comment]
Just released 2.5.2 and manually verified that your use-case works. :-)

**Comment by afranken** [comment]
Looking at your example again - are you mounting that directory into the S3Mock root to simulate already uploaded objects?

This _may_ (accidentally) have worked in 2.4 and prior, starting with 2.5, buckets and objects are managed differently - they will have to be created properly through the S3 API, otherwise the S3Mock will return 404 Not Found.

**Comment by JoeNonExotic** [comment]
`are you mounting that directory into the S3Mock root to simulate already uploaded objects?` : Yes, that was the plan initially, hence the report and it seems that its not meant to be used that way based on your [commit message](https://github.com/adobe/S3Mock/commit/626b13b096b3d82cc900f3e1368618ceb3774232) , but acknowledged that I will use API to do this instead. Thanks a lot for the quick turnaround :)

**Comment by afranken** [comment]
Just as a clarification:
If you do not delete your objects and buckets (for example in a `AfterAll` method that just cleans up everything), S3Mock will by default delete all files and directories when shutting down.
When `retainFilesOnExit` is set to `true`, S3Mock will not remove those files and directories.

I will document the directory layout properly soon (was requested a while ago in #220, but I wanted to refactor the way we store files first before doing the documentation, which I just finished in 2.5.0), but of the top of my head it should look like this:

```Storage of Buckets:
<root>/<bucket-name>

Storage of Bucket metadata:
<root>/<bucket-name>/bucketMetadata
(contains the "key" -> "uuid" dictionary for all objects uploaded to this bucket)

Storage of objects
<root>/<bucket-name>/<uuid>

Storage of object (meta)data
<root>/<bucket-name>/<uuid>/binarydata
<root>/<bucket-name>/<uuid>/metadata
<root>/<bucket-name>/<uuid>/partdata
<root>/<bucket-name>/<uuid>/<partNo>.part
```

**Comment by afranken** [comment]
finished documenting the file system structure:
https://github.com/adobe/S3Mock#file-system-structure

**Comment by masterkain** [comment]
I have an issue with 2.5.x+ both locally with docker compose and on kubernetes. I use `retainFilesOnExit: true`

I tried emptying the volumes thinking about a different filesystem structure between releases but this keeps happening on restart. Now using 2.6.1.

```
│ 2022-09-09 18:31:00.476  INFO 1 --- [           main] c.a.testing.s3mock.S3MockApplication     : Starting S3MockApplication using Java 17.0.4.1 on drone-cache-s3mock-67d655499f-gxr29 with │
│ 2022-09-09 18:31:00.479  INFO 1 --- [           main] c.a.testing.s3mock.S3MockApplication     : No active profile set, falling back to 1 default profile: "default"                        ││ 2022-09-09 18:31:01.281  INFO 1 --- [           main] org.eclipse.jetty.util.log               : Logging initialized @1574ms to org.eclipse.jetty.util.log.Slf4jLog                         │
│ 2022-09-09 18:31:01.393  INFO 1 --- [           main] o.s.b.w.e.j.JettyServletWebServerFactory : Server initialized with port: 9191                                                         ││ 2022-09-09 18:31:01.408  INFO 1 --- [           main] org.eclipse.jetty.server.Server          : jetty-9.4.48.v20220622; built: 2022-06-21T20:42:25.880Z; git: 6b67c5719d1f4371b33655ff2d04 │
│ 2022-09-09 18:31:01.437  INFO 1 --- [           main] o.e.j.s.h.ContextHandler.application     : Initializing Spring embedded WebApplicationContext                                         ││ 2022-09-09 18:31:01.437  INFO 1 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 918 ms                             │
│ 2022-09-09 18:31:01.476  WARN 1 --- [           main] i.m.c.i.binder.jvm.JvmGcMetrics          : GC notifications will not be available because com.sun.management.GarbageCollectionNotific ││ 2022-09-09 18:31:01.753  INFO 1 --- [           main] org.eclipse.jetty.server.session         : DefaultSessionIdManager workerName=node0                                                   │
│ 2022-09-09 18:31:01.753  INFO 1 --- [           main] org.eclipse.jetty.server.session         : No SessionScavenger set, using defaults                                                    ││ 2022-09-09 18:31:01.755  INFO 1 --- [           main] org.eclipse.jetty.server.session         : node0 Scavenging every 600000ms                                                            │
│ 2022-09-09 18:31:01.759  INFO 1 --- [           main] o.e.jetty.server.handler.ContextHandler  : Started o.s.b.w.e.j.JettyEmbeddedWebAppContext@6273c5a4{application,/,[file:///tmp/jetty-d ││ 2022-09-09 18:31:01.760  INFO 1 --- [           main] org.eclipse.jetty.server.Server          : Started @2054ms                                                                            │
│ 2022-09-09 18:31:01.799  INFO 1 --- [           main] c.a.t.s3mock.store.StoreConfiguration    : Using existing folder "/s3mockroot" as root folder. Will retain files on exit: true        ││ 2022-09-09 18:31:01.800  WARN 1 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.spri │
│ 2022-09-09 18:31:01.803  INFO 1 --- [           main] org.eclipse.jetty.server.session         : node0 Stopped scavenging                                                                   ││ 2022-09-09 18:31:01.804  INFO 1 --- [           main] o.e.jetty.server.handler.ContextHandler  : Stopped o.s.b.w.e.j.JettyEmbeddedWebAppContext@6273c5a4{application,/,[file:///tmp/jetty-d │
│ 2022-09-09 18:31:01.814  INFO 1 --- [           main] ConditionEvaluationReportLoggingListener :                                                                                            ││                                                                                                                                                                                             │
│ Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.                                                                           ││ 2022-09-09 18:31:01.830 ERROR 1 --- [           main] o.s.boot.SpringApplication               : Application run failed                                                                     │
│                                                                                                                                                                                             ││ org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'bucketStore' defined in class path resource [com/adobe/testing/s3mock/store/StoreConfiguration.clas │
│     at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:658) ~[spring-beans-5.3.22.jar!/:5.3.22]                                          ││     at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:638) ~[spring-beans-5.3.22.jar!/:5.3.22]                        │
│     at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352) ~[spring-beans-5.3.22.jar!/ ││     at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195) ~[spring-beans-5.3.22.jar!/:5.3.22]    │
│     at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582) ~[spring-beans-5.3.22.jar!/:5.3.22]           ││     at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542) ~[spring-beans-5.3.22.jar!/:5.3.22]             │
│     at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335) ~[spring-beans-5.3.22.jar!/:5.3.22]                                   ││     at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-5.3.22.jar!/:5.3.22]                       │
│     at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333) ~[spring-beans-5.3.22.jar!/:5.3.22]                                            ││     at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208) ~[spring-beans-5.3.22.jar!/:5.3.22]                                              │
│     at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:955) ~[spring-beans-5.3.22.jar!/:5.3.22]               ││     at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:918) ~[spring-context-5.3.22.jar!/:5.3.22]            │
│     at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583) ~[spring-context-5.3.22.jar!/:5.3.22]                                    ││     at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:147) ~[spring-boot-2.7.3.jar!/:2.7.3]                │
│     at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:734) ~[spring-boot-2.7.3.jar!/:2.7.3]                                                                      ││     at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:408) ~[spring-boot-2.7.3.jar!/:2.7.3]                                                               │
│     at org.springframework.boot.SpringApplication.run(SpringApplication.java:308) ~[spring-boot-2.7.3.jar!/:2.7.3]                                                                          ││     at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:164) ~[spring-boot-2.7.3.jar!/:2.7.3]                                                    │
│     at com.adobe.testing.s3mock.S3MockApplication.start(S3MockApplication.java:181) ~[classes!/:na]                                                                                         ││     at com.adobe.testing.s3mock.S3MockApplication.start(S3MockApplication.java:145) ~[classes!/:na]                                                                                         │
│     at com.adobe.testing.s3mock.S3MockApplication.main(S3MockApplication.java:134) ~[classes!/:na]                                                                                          ││     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]                                                                                              │
│     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[na:na]                                                                                              │
│     at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[na:na]                                                                                          │
│     at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[na:na]                                                                                                                   │
│     at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49) ~[s3mock.jar:na]                                                                                      │
│     at org.springframework.boot.loader.Launcher.launch(Launcher.java:108) ~[s3mock.jar:na]                                                                                                  │
│     at org.springframework.boot.loader.Launcher.launch(Launcher.java:58) ~[s3mock.jar:na]                                                                                                   │
│     at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:65) ~[s3mock.jar:na]                                                                                               │
│ Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.adobe.testing.s3mock.store.BucketStore]: Factory method 'bucketStore' threw exception; nested e │
│     at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185) ~[spring-beans-5.3.22.jar!/:5.3.22]                          │
│     at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653) ~[spring-beans-5.3.22.jar!/:5.3.22]                                          │
│     ... 28 common frames omitted                                                                                                                                                            │
│ Caused by: java.lang.NullPointerException: Cannot enter synchronized block because the return value of "java.util.Map.get(Object)" is null                                                  │
│     at com.adobe.testing.s3mock.store.BucketStore.getBucketMetadata(BucketStore.java:93) ~[classes!/:na]                                                                                    │
│     at com.adobe.testing.s3mock.store.BucketStore.createBucket(BucketStore.java:182) ~[classes!/:na]                                                                                        │
│     at com.adobe.testing.s3mock.store.BucketStore.lambda$new$0(BucketStore.java:64) ~[classes!/:na]                                                                                         │
│     at java.base/java.util.ArrayList.forEach(Unknown Source) ~[na:na]                                                                                                                       │
│     at com.adobe.testing.s3mock.store.BucketStore.<init>(BucketStore.java:64) ~[classes!/:na]                                                                                               │
│     at com.adobe.testing.s3mock.store.StoreConfiguration.bucketStore(StoreConfiguration.java:51) ~[classes!/:na]                                                                            │
│     at com.adobe.testing.s3mock.store.StoreConfiguration$$EnhancerBySpringCGLIB$$8317884a.CGLIB$bucketStore$1(<generated>) ~[classes!/:na]                                                  │
│     at com.adobe.testing.s3mock.store.StoreConfiguration$$EnhancerBySpringCGLIB$$8317884a$$FastClassBySpringCGLIB$$2351a42b.invoke(<generated>) ~[classes!/:na]                             │
│     at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244) ~[spring-core-5.3.22.jar!/:5.3.22]                                                                     │
│     at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331) ~[spring-context-5.3.22.jar!/:5.3.22]         │
│     at com.adobe.testing.s3mock.store.StoreConfiguration$$EnhancerBySpringCGLIB$$8317884a.bucketStore(<generated>) ~[classes!/:na]                                                          │
│     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]                                                                                              │
│     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[na:na]                                                                                              │
│     at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[na:na]                                                                                          │
│     at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[na:na]                                                                                                                   │
│     at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[spring-beans-5.3.22.jar!/:5.3.22]                          │
│     ... 29 common frames omitted
```

**Comment by afranken** [comment]
@masterkain 
So you are starting S3Mock with a mounted volume and `retainFilesOnExit: true`, then restart it on that volume?
See:
https://github.com/adobe/S3Mock#file-system-structure
`retainFilesOnExit` is meant to help with debugging, not meant to pre-populate the S3 buckets you are using in your tests.

This is not expected to work (although this may have worked up to 2.4.x for some use-cases) - S3Mock expects the root directory to be empty when starting, and not all data necessary to serve requests is persisted to disk.
The format / file names used to store data locally is also not part of the public API of the S3Mock, and may change even in patch version releases.

**Comment by masterkain** [comment]
thanks for the reply, but I guess don't understand the point of the `retainFilesOnExit` option.

I have some use cases for s3mock, I use it in specs, as a local s3 compatible container for storing development uploads and on kubernetes as an s3 layer to store cached data between builds.

On kubernetes I have it on auto image update and `retainFilesOnExit`, the problem did start with the 2.5.x series I believe.
So it's not data that I should persist long term but I kinda expect to have `retainFilesOnExit` at least not to crash on startup. I might even be fine with filesystem getting wiped on every version update, but at least not crash this way especially given I had no troubles in the past.

Based on what I gather here I should find another s3 compatible layer that properly persist data between restarts, any suggestions?

**Comment by afranken** [comment]
@masterkain 
well, for now, you can just stay on 2.4.x unless you desperately need features added in 2.5 and later. :)

> So it's not data that I should persist long term but I kinda expect to have retainFilesOnExit at least not to crash on startup. I might even be fine with filesystem getting wiped on every version update, but at least not crash this way especially given I had no troubles in the past (as long as at least with the same version something is persisted between restarts).

Let me think about this, maybe there is a low-maintenance way to support starting the S3Mock if the file system was created during an earlier run of the same version...
